### PR TITLE
Enable devicesAndModulesInTargetDeviceScope REST endpoint

### PR DESCRIPTION
**Pull request by richma-ms**
This change adds a new DeviceScopeController to service the devicesAndModulesInTargetDeviceScope REST endpoint in EdgeHub, so that child Edge devices can call their parent EdgeHub to retrieve scope information.

### CODE REVIEW

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/NestedDeviceScopeApiClient.cs [hunk 1]
```diff
@@ -140,6 +140,7 @@ async Task<ScopeResult> GetIdentitiesInTargetScope(Uri uri, Option<string> conti
                 var payload = new NestedScopeRequest(this.batchSize, continuationToken.OrDefault(), authChain);
                 string token = await this.edgeHubTokenProvider.GetTokenAsync(Option.None<TimeSpan>());
                 msg.Headers.Add(HttpRequestHeader.Authorization.ToString(), token);
+                msg.Headers.Add(Constants.ServiceApiIdHeaderKey, this.actorEdgeDeviceId + "/" + this.moduleId);
```
Comment (varunpuranik) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/NestedDeviceScopeApiClient.cs:143]: Why this change?

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/EdgeHubScopeDevice.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Hub.Http.Controllers
+{
+    using System.Collections.Generic;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Azure.Devices.Shared;
+    using Newtonsoft.Json;
+    using Newtonsoft.Json.Converters;
+
+    public class EdgeHubScopeDevice
```
Comment (varunpuranik) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/EdgeHubScopeDevice.cs:10]: What is an EdgeHubScopeDevice? Can you pls add a comment?

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/Microsoft.Azure.Devices.Edge.Hub.Http.csproj [hunk 3]
```diff
@@ -19,6 +19,7 @@
   <ItemGroup>
     <ProjectReference Include="..\..\..\..\edge-util\src\Microsoft.Azure.Devices.Edge.Util\Microsoft.Azure.Devices.Edge.Util.csproj" />
     <ProjectReference Include="..\Microsoft.Azure.Devices.Edge.Hub.Core\Microsoft.Azure.Devices.Edge.Hub.Core.csproj" />
+    <ProjectReference Include="..\Microsoft.Azure.Devices.Edge.Hub.CloudProxy\Microsoft.Azure.Devices.Edge.Hub.CloudProxy.csproj" />
```
Comment (varunpuranik) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/Microsoft.Azure.Devices.Edge.Hub.Http.csproj:22]: Let's avoid this dependency. We can move the required types to core.

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        readonly Task<IEdgeHub> edgeHubGetter;
+
+        public DeviceScopeController(Task<IEdgeHub> edgeHub)
+        {
+            this.edgeHubGetter = Preconditions.CheckNotNull(edgeHub, nameof(edgeHub));
+        }
+
+        [HttpPost]
+        [Route("devices/{actorDeviceId}/modules/{actorModuleId}/devicesAndModulesInTargetDeviceScope")]
+        public async Task InvokeAsync([FromRoute] string actorDeviceId, [FromRoute] string actorModuleId, [FromBody] NestedScopeRequest request)
```
Comment (varunpuranik) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs:34]: Nit: Can we name it more specific - like GetDevicesAndModulesInTargetScope or so?

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        {
+            actorDeviceId = WebUtility.UrlDecode(Preconditions.CheckNonWhiteSpace(actorDeviceId, nameof(actorDeviceId)));
+            actorModuleId = WebUtility.UrlDecode(Preconditions.CheckNonWhiteSpace(actorModuleId, nameof(actorModuleId)));
+            Preconditions.CheckArgument(actorModuleId == Constants.EdgeHubModuleId);
+
+            EdgeHubScopeResult result = await this.HandleDevicesAndModulesInTargetDeviceScope(actorDeviceId, actorModuleId, request);
+            await this.SendResponse(result);
+        }
+
+        public async Task<EdgeHubScopeResult> HandleDevicesAndModulesInTargetDeviceScope(string actorDeviceId, string actorModuleId, NestedScopeRequest request)
```
Comment (varunpuranik) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs:44]: Nit: private?

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/EdgeHubScopeResult.cs [hunk 6]
```diff
@@ -0,0 +1,23 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Hub.CloudProxy
+{
+    using System.Collections.Generic;
+    using System.Net;
+    using Microsoft.Azure.Devices.Edge.Hub.Http.Controllers;
+    using Newtonsoft.Json;
+
+    public class EdgeHubScopeResult
+    {
```
Comment (varunpuranik) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/EdgeHubScopeResult.cs:10]: Pls make this immutable

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/EdgeHubScopeDevice.cs [hunk 7]
```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Hub.Http.Controllers
```
Comment (varunpuranik) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/EdgeHubScopeDevice.cs:2]: Nit: remove .Controllers

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/EdgeHubScopeDevice.cs (hunk 2 above)
Comment (varunpuranik) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/EdgeHubScopeDevice.cs:10]: Nit: pls make this immutable.

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+                var resultUtf8Bytes = Encoding.UTF8.GetBytes(resultJsonContent);
+
+                this.Response.ContentLength = resultUtf8Bytes.Length;
+                this.Response.ContentType = SupportedContentType;
+
+                await this.Response.Body.WriteAsync(resultUtf8Bytes, 0, resultUtf8Bytes.Length);
+            }
+        }
+
+        public static bool ValidateChainAndGetTargetDeviceId(string actorDeviceId, string authChain, out string targetDeviceId)
```
Comment (varunpuranik) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs:97]: internal

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+        {
+            targetDeviceId = string.Empty;
+            var actorAuthChainIds = authChain.Split(';', StringSplitOptions.RemoveEmptyEntries).ToList();
```
Comment (varunpuranik) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs:100]: nit: why toList?

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+            // The target device ID is the first device ID in the chain
+            if (!actorAuthChainIds[0].Contains('/'))
+            {
+                targetDeviceId = actorAuthChainIds[0];
+            }
+            else
+            {
+                // The first element is a module ID, so the target device
+                // should be the second element in the chain
+                if (actorAuthChainIds.Count < 2 || actorAuthChainIds[1].Contains('/'))
```
Comment (varunpuranik) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs:124]: Should we instead get the "device" that the module belongs to?

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/middleware/AuthenticationMiddleware.cs [hunk 11]
```diff
@@ -133,12 +133,6 @@ internal async Task<(bool, string)> AuthenticateRequest(HttpContext context)
             IIdentity identity = clientCredentials.Identity;
             IAuthenticator authenticator = await this.authenticatorTask;
 
-            // A downstream module cannot invoke a method on a device by connecting to EdgeHub
-            if (deviceId != this.edgeDeviceId)
```
Comment (varunpuranik) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/middleware/AuthenticationMiddleware.cs:137]: I think this should move to the TwinsController?

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/middleware/AuthenticationMiddleware.cs [hunk 12]
```diff
@@ -63,7 +63,7 @@ public async Task Invoke(HttpContext context)
 
         internal async Task<(bool, string)> AuthenticateRequest(HttpContext context)
         {
-            if (!context.Request.Headers.TryGetValue(HttpConstants.IdHeaderKey, out StringValues clientIds) || clientIds.Count == 0)
+            if (!context.Request.Headers.TryGetValue(Constants.ServiceApiIdHeaderKey, out StringValues clientIds) || clientIds.Count == 0)
```
Comment (varunpuranik) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/middleware/AuthenticationMiddleware.cs:66]: I think we should do authn/authz on a per controller basis instead of in a centralized location like Middleware

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.Http.Test/AuthenticationMiddlewareTest.cs [hunk 13]
```diff
@@ -328,30 +304,6 @@ public async Task InvalidCredentialsRequestX509Test_AuthFailed()
             Assert.Equal("Unable to authenticate device with Id device_2/module_1", result.message);
         }
 
-        [Fact]
-        public async Task InvalidAuthenticateRequestX509Test_InvalidDeviceId()
-        {
```
Comment (richma-ms) [inline edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.Http.Test/AuthenticationMiddlewareTest.cs:333]: Varun: don't remove the tests, just disable for now until we have the authentication middleware doing the right thing with multiple service APIs.

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/storage/MessageStore.cs [hunk 14]
```diff
@@ -446,12 +446,15 @@ internal static void MessageNotFound(string edgeMessageId)
...
+                // Log.LogDebug((int)EventIds.GettingNextBatch, $"Getting next batch for endpoint {entityName} starting from {startingOffset} with batch size {batchSize}.");
             }
 
             internal static void ObtainedNextBatch(string entityName, long startingOffset, int count)
             {
-                Log.LogDebug((int)EventIds.ObtainedNextBatch, $"Obtained next batch for endpoint {entityName} with batch size {count}. Next start offset = {startingOffset}.");
+                if (count > 0)
+                {
+                    Log.LogDebug((int)EventIds.ObtainedNextBatch, $"Obtained next batch for endpoint {entityName} with batch size {count}. Next start offset = {startingOffset}.");
+                }
```
Comment (richma-ms) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/storage/MessageStore.cs:457]: need to revert, was used only for local testing

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/NestedDeviceScopeApiClient.cs (hunk 1 above)
Comment (philipktlin) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/NestedDeviceScopeApiClient.cs:143]: Should we add this header only if nested edge is enabled?

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/ServiceIdentityTree.cs [hunk 15]
```diff
@@ -118,6 +118,27 @@ public async Task<Option<string>> GetAuthChain(string id)
...
+            Preconditions.CheckNonWhiteSpace(id, nameof(id));
+
+            var children = new List<ServiceIdentity>();
+
+            using (await this.nodesLock.LockAsync())
+            {
+                if (this.nodes.TryGetValue(id, out ServiceIdentityTreeNode treeNode))
+                {
+                    IList<ServiceIdentityTreeNode> childNodes = treeNode.GetAllChildren();
+                    foreach (ServiceIdentityTreeNode node in childNodes)
```
Comment (philipktlin) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/ServiceIdentityTree.cs:132]: use List AddRange method.
```suggestion
                    children.AddRange(treeNode.GetAllChildren());
```

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/EdgeHubScopeDevice.cs [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+            DeviceCapabilities capabilities,
+            string deviceScope,
+            IEnumerable<string> parentScopes)
+        {
+            this.Id = Preconditions.CheckNonWhiteSpace(deviceId, nameof(deviceId));
+            this.GenerationId = Preconditions.CheckNonWhiteSpace(generationId, nameof(generationId));
+            this.Status = Preconditions.CheckNotNull(status);
+            this.Authentication = Preconditions.CheckNotNull(authentication);
+            this.Capabilities = Preconditions.CheckNotNull(capabilities);
+            this.Scope = deviceScope;
```
Comment (philipktlin) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/EdgeHubScopeDevice.cs:26]: nit: check for not null

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/EdgeHubScopeDevice.cs [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+            this.ParentScopes = parentScopes;
```
Comment (philipktlin) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/EdgeHubScopeDevice.cs:27]: is this allowed to be empty?

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/EdgeHubScopeDevice.cs [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+        public string GenerationId { get; set; }
+
+        [JsonConverter(typeof(StringEnumConverter))]
+        [JsonProperty(PropertyName = "status")]
+        public DeviceStatus Status { get; set; }
+
+        [JsonProperty(PropertyName = "authentication")]
+        public AuthenticationMechanism Authentication { get; set; }
+
+        [JsonProperty(PropertyName = "capabilities", NullValueHandling = NullValueHandling.Ignore)]
```
Comment (philipktlin) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/EdgeHubScopeDevice.cs:43]: nit: don't need NullValueHandling as it is not allowed to be null in constructor.

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/EdgeHubScopeModule.cs [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+        public string Id { get; }
+
+        [JsonProperty(PropertyName = "deviceId")]
+        public string DeviceId { get; }
+
+        [JsonProperty(PropertyName = "generationId")]
+        public string GenerationId { get; }
+
+        [JsonProperty(PropertyName = "authentication")]
+        public AuthenticationMechanism Authentication { get; set; }
```
Comment (philipktlin) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/EdgeHubScopeModule.cs:34]: nit: remove set

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs [hunk 20]
```diff
@@ -0,0 +1,191 @@
...
+            }
+
+            // Actor device is claiming to be our child, and that the target device is its child.
+            // So we should have an authchain already cached for the target device.
+            IEdgeHub edgeHub = await this.edgeHubGetter;
+            Option<string> targetAuthChainOption = await edgeHub.GetAuthChainForIdentity(targetDeviceId);
+            string targetAuthChain = targetAuthChainOption.Expect(() => new UnauthorizedAccessException($"{targetDeviceId} is not a child of this Edge device"));
+
+            // The actor device should be somewhere in the target device's authchain
+            if (!targetAuthChain.Contains(actorDeviceId))
```
Comment (philipktlin) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs:65]: What happens if target auth chain contains a partial match of actor device id, e.g. auth chain is ABC;DEF;GHI and actor device id is DE.  I think we should append delimiter at the beginning and end of auth chain and use contain with ";{actor device id};" value.

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/NestedDeviceScopeApiClient.cs (hunk 1 above)
Comment (richma-ms) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/NestedDeviceScopeApiClient.cs:143]: this header is ignored by IoT Hub, so it's safe to tack on.

---
In reply to: [451893193](https://github.com/Azure/iotedge/pull/3149#discussion_r451893193) [](ancestors = 451893193)

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/ServiceIdentityTree.cs (hunk 15 above)
Comment (richma-ms) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/ServiceIdentityTree.cs:132]: they're different types, but AddRange still works if I apply a Select

---
In reply to: [451895233](https://github.com/Azure/iotedge/pull/3149#discussion_r451895233) [](ancestors = 451895233)

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/EdgeHubScopeDevice.cs (hunk 2 above)
Comment (richma-ms) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/EdgeHubScopeDevice.cs:10]: Filed bug to track this

[https://msazure.visualstudio.com/One/_workitems/edit/7649919](https://msazure.visualstudio.com/One/_workitems/edit/7649919)

---
In reply to: [451638847](https://github.com/Azure/iotedge/pull/3149#discussion_r451638847) [](ancestors = 451638847)

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/EdgeHubScopeResult.cs (hunk 6 above)
Comment (richma-ms) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/EdgeHubScopeResult.cs:10]: the DeviceController is setting the Status, Devices, and Modules separately

---
In reply to: [451646490](https://github.com/Azure/iotedge/pull/3149#discussion_r451646490) [](ancestors = 451646490)

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs (hunk 10 above)
Comment (richma-ms) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs:124]: yes, that's safer than what I'm currnetly doing

---
In reply to: [451661275](https://github.com/Azure/iotedge/pull/3149#discussion_r451661275) [](ancestors = 451661275)

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs (hunk 20 above)
Comment (richma-ms) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs:65]: good point, I think it's probably easier to read/maintain if I just tokenize the string and do a comparison against each element.

---
In reply to: [451899836](https://github.com/Azure/iotedge/pull/3149#discussion_r451899836) [](ancestors = 451899836)

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/middleware/AuthenticationMiddleware.cs (hunk 11 above)
Comment (richma-ms) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/middleware/AuthenticationMiddleware.cs:137]: Filed bug to track refactor the middleware:

[https://msazure.visualstudio.com/One/_workitems/edit/7649811](https://msazure.visualstudio.com/One/_workitems/edit/7649811)

---
In reply to: [451716961](https://github.com/Azure/iotedge/pull/3149#discussion_r451716961) [](ancestors = 451716961)

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/middleware/AuthenticationMiddleware.cs (hunk 12 above)
Comment (richma-ms) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/middleware/AuthenticationMiddleware.cs:66]: [https://msazure.visualstudio.com/One/_workitems/edit/7649811](https://msazure.visualstudio.com/One/_workitems/edit/7649811)

---
In reply to: [451721236](https://github.com/Azure/iotedge/pull/3149#discussion_r451721236) [](ancestors = 451721236)

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs (hunk 9 above)
Comment (richma-ms) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs:100]: I think I was using something from IList in an earlier iteration, but looks like I don't have it anymore.  Changed back to array.

---
In reply to: [451660023](https://github.com/Azure/iotedge/pull/3149#discussion_r451660023) [](ancestors = 451660023)

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/EdgeHubScopeModule.cs [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+
+    public class EdgeHubScopeModule
+    {
+        public EdgeHubScopeModule(
+            string moduleId,
+            string deviceId,
+            string generationId,
+            AuthenticationMechanism authentication)
+        {
+            this.Id = Preconditions.CheckNonWhiteSpace(moduleId, nameof(deviceId));
```
Comment (ancaantochi) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/EdgeHubScopeModule.cs:18]: nameof(moduleId)

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs [hunk 22]
```diff
@@ -0,0 +1,186 @@
...
+        {
+            Events.ReceivedScopeRequest(actorDeviceId, request);
+
+            EdgeHubScopeResult result = new EdgeHubScopeResult();
+
+            // Parse the target device ID from the authchain
+            if (!ValidateChainAndGetTargetDeviceId(actorDeviceId, request.AuthChain, out string targetDeviceId))
+            {
+                Events.InvalidAuthchain(request.AuthChain);
+                result.Status = HttpStatusCode.BadRequest;
```
Comment (ancaantochi) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs:51]: shouldn't this be unauthorized?

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs [hunk 23]
```diff
@@ -0,0 +1,186 @@
...
+            this.edgeHubGetter = Preconditions.CheckNotNull(edgeHub, nameof(edgeHub));
+        }
+
+        [HttpPost]
+        [Route("devices/{actorDeviceId}/modules/{actorModuleId}/devicesAndModulesInTargetDeviceScope")]
+        public async Task GetDevicesAndModulesInTargetDeviceScope([FromRoute] string actorDeviceId, [FromRoute] string actorModuleId, [FromBody] NestedScopeRequest request)
+        {
+            actorDeviceId = WebUtility.UrlDecode(Preconditions.CheckNonWhiteSpace(actorDeviceId, nameof(actorDeviceId)));
+            actorModuleId = WebUtility.UrlDecode(Preconditions.CheckNonWhiteSpace(actorModuleId, nameof(actorModuleId)));
+            Preconditions.CheckArgument(actorModuleId == Constants.EdgeHubModuleId);
```
Comment (ancaantochi) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs:35]: unauthorized?
